src/validation.c

```diff
@@ -0,0 +1,223 @@
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 3.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <glib.h>
+#include <glib/gstdio.h>
+#include <gio/gio.h>
+#include <arpa/inet.h>
+
+#include <yaml.h>
+
+#include "parse.h"
+#include "error.h"
+
+
+/* Check sanity for address types */
+
+gboolean
+is_ip4_address(const char* address)
+{
+    struct in_addr a4;
+    int ret;
+
+    ret = inet_pton(AF_INET, address, &a4);
+    g_assert(ret >= 0);
+    if (ret > 0)
+        return TRUE;
+
+    return FALSE;
+}
+
+gboolean
+is_ip6_address(const char* address)
+{
+    struct in6_addr a6;
+    int ret;
+
+    ret = inet_pton(AF_INET6, address, &a6);
+    g_assert(ret >= 0);
+    if (ret > 0)
+        return TRUE;
+
+    return FALSE;
+}
+
+/************************************************
+ * Validation for grammar and backend rules.
+ ************************************************/
+static gboolean
+validate_tunnel_grammar(net_definition* nd, yaml_node_t* node, GError** error)
+{
+    if (nd->tunnel.mode == TUNNEL_MODE_UNKNOWN)
+        return yaml_error(node, error, "%s: missing 'mode' property for tunnel", nd->id);
+
+    /* Validate local/remote IPs */
+    if (!nd->tunnel.local_ip)
+        return yaml_error(node, error, "%s: missing 'local' property for tunnel", nd->id);
+    if (!nd->tunnel.remote_ip)
+        return yaml_error(node, error, "%s: missing 'remote' property for tunnel", nd->id);
+
+    switch(nd->tunnel.mode) {
+        case TUNNEL_MODE_IPIP6:
+        case TUNNEL_MODE_IP6IP6:
+        case TUNNEL_MODE_IP6GRE:
+        case TUNNEL_MODE_IP6GRETAP:
+        case TUNNEL_MODE_VTI6:
+            if (!is_ip6_address(nd->tunnel.local_ip))
+                return yaml_error(node, error, "%s: 'local' must be a valid IPv6 address for this tunnel type", nd->id);
+            if (!is_ip6_address(nd->tunnel.remote_ip))
+                return yaml_error(node, error, "%s: 'remote' must be a valid IPv6 address for this tunnel type", nd->id);
+            break;
+
+        default:
+            if (!is_ip4_address(nd->tunnel.local_ip))
+                return yaml_error(node, error, "%s: 'local' must be a valid IPv4 address for this tunnel type", nd->id);
+            if (!is_ip4_address(nd->tunnel.remote_ip))
+                return yaml_error(node, error, "%s: 'remote' must be a valid IPv4 address for this tunnel type", nd->id);
+            break;
+    }
+
+    return TRUE;
+}
+
+static gboolean
+validate_tunnel_backend_rules(net_definition* nd, yaml_node_t* node, GError** error)
```

Comment:
I don't think node here can ever be non-NULL can it? I might be missing something.